File: frontend/mobile_app/lib/data/auth_repository.dart
import 'dart:convert';

import 'package:http/http.dart' as http;

import '../core/config/app_config.dart';
import '../models/session.dart';

/// A thrown auth failure carrying a human-readable message from the backend
/// (e.g. "That username is already taken", "Invalid username or password").
class AuthException implements Exception {
  const AuthException(this.message);
  final String message;
  @override
  String toString() => message;
}

/// The result of a successful register/login: the signed-in associate plus the
/// bearer token used to authenticate subsequent calls (e.g. checkout).
class AuthResult {
  const AuthResult({required this.salesperson, required this.token});

  factory AuthResult.fromJson(Map<String, dynamic> json) => AuthResult(
    salesperson:
        Salesperson.fromJson(json['salesperson'] as Map<String, dynamic>),
    token: json['token'] as String,
  );

  final Salesperson salesperson;
  final String token;
}

/// Salesperson authentication. Swapped for [MockAuthRepository] in standalone
/// mode so the offline demo still works without a backend.
abstract interface class AuthRepository {
  Future<AuthResult> register({
    required String name,
    String? title,
    required String username,
    required String password,
  });

  Future<AuthResult> login({
    required String username,
    required String password,
  });

  /// Best-effort server-side token revocation. Never throws.
  Future<void> logout(String token);
}

/// [AuthRepository] backed by the Node backend (`/api/auth/*`).
class HttpAuthRepository implements AuthRepository {
  HttpAuthRepository({http.Client? client})
    : _client = client ?? http.Client();

  final http.Client _client;
  static const Duration _timeout = Duration(seconds: 8);

  @override
  Future<AuthResult> register({
    required String name,
    String? title,
    required String username,
    required String password,
  }) => _post('/api/auth/register', <String, dynamic>{
    'name': name,
    if (title != null && title.trim().isNotEmpty) 'title': title.trim(),
    'username': username,
    'password': password,
  });

  @override
  Future<AuthResult> login({
    required String username,
    required String password,
  }) => _post('/api/auth/login', <String, dynamic>{
    'username': username,
    'password': password,
  });

  @override
  Future<void> logout(String token) async {
    try {
      await _client
          .post(
            AppConfig.http('/api/auth/logout'),
            headers: <String, String>{'authorization': 'Bearer $token'},
          )
          .timeout(_timeout);
    } catch (_) {
      // Logout is best-effort; the client clears local state regardless.
    }
  }

  Future<AuthResult> _post(String path, Map<String, dynamic> body) async {
    late final http.Response res;
    try {
      res = await _client
          .post(
            AppConfig.http(path),
            headers: <String, String>{'content-type': 'application/json'},
            body: jsonEncode(body),
          )
          .timeout(_timeout);
    } catch (_) {
      throw const AuthException('Could not reach the server. Check the connection.');
    }
    final Map<String, dynamic> json =
        jsonDecode(res.body) as Map<String, dynamic>;
    if (res.statusCode == 200 || res.statusCode == 201) {
      return AuthResult.fromJson(json);
    }
    final Object? err = json['error'];
    final String message = err is Map<String, dynamic>
        ? (err['message'] as String? ?? 'Authentication failed')
        : 'Authentication failed';
    throw AuthException(message);
  }
}

/// Offline/demo auth: accepts any credentials and returns a synthetic associate.
class MockAuthRepository implements AuthRepository {
  const MockAuthRepository();

  @override
  Future<AuthResult> register({
    required String name,
    String? title,
    required String username,
    required String password,
  }) async => AuthResult(
    salesperson: Salesperson(id: 'mock_$username', name: name, title: title),
    token: 'mock-token-$username',
  );

  @override
  Future<AuthResult> login({
    required String username,
    required String password,
  }) async => AuthResult(
    salesperson: Salesperson(
      id: 'mock_$username',
      name: username.isEmpty ? 'Associate' : username,
      title: 'Studio Associate',
    ),
    token: 'mock-token-$username',
  );

  @override
  Future<void> logout(String token) async {}
}

File: frontend-rn/mobile_app/src/data/authRepository.ts
import { AppConfig } from '../config/appConfig';
import { Salesperson } from '../models/session';

/// A thrown auth failure carrying a human-readable message from the backend.
export class AuthException extends Error {}

/// The result of a successful register/login.
export class AuthResult {
  readonly salesperson: Salesperson;
  readonly token: string;
  constructor(init: { salesperson: Salesperson; token: string }) {
    this.salesperson = init.salesperson;
    this.token = init.token;
  }
  static fromJson(json: any): AuthResult {
    return new AuthResult({
      salesperson: Salesperson.fromJson(json.salesperson),
      token: json.token as string,
    });
  }
}

export interface AuthRepository {
  register(opts: { name: string; title?: string; username: string; password: string }): Promise<AuthResult>;
  login(opts: { username: string; password: string }): Promise<AuthResult>;
  /// Best-effort server-side token revocation. Never throws.
  logout(token: string): Promise<void>;
}

/// AuthRepository backed by the Node backend (`/api/auth/*`). Ported 1:1.
export class HttpAuthRepository implements AuthRepository {
  private static readonly timeoutMs = 8000;

  async register(opts: { name: string; title?: string; username: string; password: string }): Promise<AuthResult> {
    const body: Record<string, any> = { name: opts.name, username: opts.username, password: opts.password };
    if (opts.title != null && opts.title.trim().length > 0) body.title = opts.title.trim();
    return this.post('/api/auth/register', body);
  }

  login(opts: { username: string; password: string }): Promise<AuthResult> {
    return this.post('/api/auth/login', { username: opts.username, password: opts.password });
  }

  async logout(token: string): Promise<void> {
    try {
      await this.fetchTimeout(AppConfig.http('/api/auth/logout'), {
        method: 'POST',
        headers: { authorization: `Bearer ${token}` },
      });
    } catch {
      // Logout is best-effort; the client clears local state regardless.
    }
  }

  private async fetchTimeout(url: string, init: RequestInit): Promise<Response> {
    const controller = new AbortController();
    const t = setTimeout(() => controller.abort(), HttpAuthRepository.timeoutMs);
    try {
      return await fetch(url, { ...init, signal: controller.signal });
    } finally {
      clearTimeout(t);
    }
  }

  private async post(path: string, body: Record<string, any>): Promise<AuthResult> {
    let res: Response;
    try {
      res = await this.fetchTimeout(AppConfig.http(path), {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(body),
      });
    } catch {
      throw new AuthException('Could not reach the server. Check the connection.');
    }
    const json = await res.json().catch(() => ({}));
    if (res.status === 200 || res.status === 201) {
      return AuthResult.fromJson(json);
    }
    const err = json.error;
    const message =
      err != null && typeof err === 'object' ? (err.message as string) ?? 'Authentication failed' : 'Authentication failed';
    throw new AuthException(message);
  }
}

/// Offline/demo auth: accepts any credentials and returns a synthetic associate.
export class MockAuthRepository implements AuthRepository {
  async register(opts: { name: string; title?: string; username: string; password: string }): Promise<AuthResult> {
    return new AuthResult({
      salesperson: new Salesperson({ id: `mock_${opts.username}`, name: opts.name, title: opts.title, username: opts.username }),
      token: `mock-token-${opts.username}`,
    });
  }

  async login(opts: { username: string; password: string }): Promise<AuthResult> {
    return new AuthResult({
      salesperson: new Salesperson({
        id: `mock_${opts.username}`,
        name: opts.username.length === 0 ? 'Associate' : opts.username,
        title: 'Studio Associate',
        username: opts.username,
      }),
      token: `mock-token-${opts.username}`,
    });
  }

  async logout(): Promise<void> {}
}
